File: 0234-palindrome-linked-list/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode h1 = head;
        int count = 0;
        while (h1 != null) { // counting nodes
            h1 = h1.next;
            count++;
        }
        ListNode left = null;
        ListNode right = null;
        ListNode prev = null;
        // splitting by the middle and reversing first half
        for (int i = 1; i <= count / 2; i++) {
            right = head.next;
            head.next = prev;
            prev = head;
            left = head;
            head = right;
            if (i == count / 2 && count % 2 == 1) { // handle right half start for odd node count
                right = right.next;
            }
        }

        // comparing to half
        while (left != null) {
            if (left.val != right.val) { // if value not equal -> return false
                return false;
            }
            left = left.next;
            right = right.next;
        }

        return true; // return true if all values was equal
    }
}

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        curr_node = head
        stack = []
        while curr_node is not None:
            stack.append(curr_node.val) # 1->2->2->1
            curr_node = curr_node.next
        while head is not None:
            c = stack.pop()
            if c != head.val:
                return False
            head = head.next
        return True
    


        